File: src/leetcode/19. Remove Nth Node From End of List/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int count = 0;
        ListNode temp = head;
        while(temp != null) {
            temp = temp.next;
            count++;
        }

        int index = count - n;

        ListNode node = new ListNode();
        node.next = head;

        ListNode tmp = node;
        for(int i = 0; i < count; i++) {
            if(i == index) {
                tmp.next = tmp.next != null ? tmp.next.next : null;
            } else {
                tmp = tmp.next;
            }                 
        }

        return node.next;
    }
}

File: src/leetcode/19. Remove Nth Node From End of List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode();
        dummy->next = head;

        ListNode* fast = head;

        int length = 0;
        while(fast != NULL) {
            length++;
            fast = fast->next;
        }

        length -= n;
        fast = dummy;
        while(length > 0) {            
            fast = fast->next;
            length--;
        }

        fast->next = fast->next->next;
        

        return dummy->next;
    }
};
